network/gossip.go

```diff
@@ -53,32 +58,62 @@ func (t *Topic) Subscribe(handler func(obj interface{})) error {
 	return nil
 }
 
+func (t *Topic) Close() error {
+	close(t.unsubscribeCh)
+	t.wg.Wait()
+
+	return t.topic.Close()
+}
+
 func (t *Topic) readLoop(sub *pubsub.Subscription, handler func(obj interface{})) {
 	ctx, cancelFn := context.WithCancel(context.Background())
 
+	var unsubscribe int32 = 0
+
+	workqueue := make(chan proto.Message, runtime.NumCPU())
+
+	t.wg.Add(1)
+
 	go func() {
-		<-t.closeCh
+		<-t.unsubscribeCh
+		atomic.StoreInt32(&unsubscribe, 1)
+
+		close(workqueue)
+
 		cancelFn()
+		sub.Cancel()
+		t.wg.Done()
 	}()
 
-	for {
+	for i := 0; i < runtime.NumCPU(); i++ {
+		go func() {
+			for {
+				obj, ok := <-workqueue
+				if !ok {
+					return
+				}
+
+				handler(obj)
+			}
+		}()
+	}
+
+	for atomic.LoadInt32(&unsubscribe) == 0 {
 		msg, err := sub.Next(ctx)
 		if err != nil {
 			t.logger.Error("failed to get topic", "err", err)
 
 			continue
 		}
 
-		go func() {
-			obj := t.createObj()
-			if err := proto.Unmarshal(msg.Data, obj); err != nil {
-				t.logger.Error("failed to unmarshal topic", "err", err)
+		obj := t.createObj()
+		if err := proto.Unmarshal(msg.Data, obj); err != nil {
+			t.logger.Error("failed to unmarshal topic", "err", err)
 
-				return
-			}
+			return
```

Comment:
What if the message is coming from an attacker?
Refactor should consider more business over pattern.

Reply:
If the message comes from an attacker, creating goroutine will run out of memory (not waiting message handler, not drop message, buffer size limit invalid [gossip.go#L19](https://github.com/dogechain-lab/dogechain/blob/c6944f2038cf974cc9c58f5f7efc3ccc1b306461/network/gossip.go#L19))

Reply:
e.g. 
An attacker send a not known type message, all nodes would return and do not receive any message any more?

Reply:
fcfcbc4 I missed this, I should use continue